### ISSUE TITLE
TODO 38: fp16-safe clamp threshold in ml_engine/utils/box_ops.py

### ISSUE THREAD

**Issue by hongthepotato** [hongthepotato/Grounded-Segment-Anything#86]
## Summary

[`ml_engine/utils/box_ops.py`](https://github.com/hongthepotato/Grounded-Segment-Anything/blob/agentic/ml_engine/utils/box_ops.py) uses `union.clamp(min=1e-12)` and `enclosing.clamp(min=1e-12)` as defense in depth against degenerate (zero-area) boxes. **`1e-12` underflows to 0 in fp16**, making the clamp a no-op under fp16 autocast. Currently masked because `loss_boxes` and `HungarianMatcher.forward` mix fp16 pred boxes with fp32 target boxes — PyTorch promotes the IoU computation to fp32 where 1e-12 is representable. A fully-fp16 pipeline would re-expose 0/0 NaN risk.

**Priority:** P3 (latent, currently shielded; only triggers if someone calls `.half()` on targets too).

## Repro

```python
>>> torch.tensor(1e-12, dtype=torch.float16).item()
0.0
```

## Why

Currently `criterion(...)` works under fp16 autocast because:
- `pred_boxes` arrive in fp16 (from autocast)
- `target_boxes` arrive in fp32 (from CPU dataloader)
- `torch.max(fp16, fp32)` → fp32 (PyTorch type promotion)
- IoU computation runs in fp32 where `clamp(min=1e-12)` is effective

If someone explicitly casts targets to fp16 (e.g., calls `.half()` on the dataloader output for a fully-fp16 training loop), the IoU computation stays in fp16, the clamp underflows to 0, and a degenerate predicted box (`w` or `h` underflowed from `sigmoid(-20)` to 0) produces `0/0 = NaN` directly.

## Tradeoff investigated and rejected

Bumping the threshold to `1e-6` is fp16-safe (1e-6 is representable in fp16, ≈9.5e-7 nearest) but **reintroduces the exact bias the PR was fixing** for area<1e-6 boxes. Two unit tests in `tests/unit/test_losses.py::TestLocalBoxIoU` directly fail with that threshold:
- `test_self_match_iou_exactly_one_for_microscopic_box` (side=1e-4, area=1e-8)
- `test_self_match_giou_exactly_one_across_box_sizes` at side=1e-4

So `1e-6` is a strict regression on the PR that introduced this code.

## Better fix candidates

**Option A — dtype-aware clamp (recommended):**
```python
iou = inter / union.clamp(min=torch.finfo(union.dtype).tiny)
# fp16 → 6.10e-5  (acceptable; biases boxes with area<6e-5 in fp16 only,
#                  which already underflow upstream anyway)
# fp32 → 1.18e-38 (effectively 0 floor — no bias)
# fp64 → 2.22e-308
```

**Option B — explicit dtype promotion:**
```python
# Always compute IoU in fp32, return in input dtype
iou = (inter.float() / union.float().clamp(min=1e-12)).to(inter.dtype)
```

Option A is cleaner. Option B works but is more intrusive.

## Acceptance criteria

- [ ] `_box_iou` and `_generalized_box_iou` work correctly under all of {fp16, fp32, fp64}
- [ ] Existing `test_self_match_iou_exactly_one_for_microscopic_box` still passes (no fp32 bias regression)
- [ ] New parametrized unit test fixture covers fp16 path explicitly:
  ```python
  @pytest.mark.parametrize("dtype", [torch.float16, torch.float32, torch.float64])
  def test_box_iou_dtype_safe_no_nan_on_degenerate(dtype):
      point = torch.tensor([[0.5, 0.5, 0.5, 0.5]], dtype=dtype)
      iou, _ = box_iou(point, point)
      assert not torch.isnan(iou).any()
      assert not torch.isinf(iou).any()
  ```
- [ ] Full suite still 2348 passing

## Surfaced via

Adversarial review during `/ship` of branch `integration-tests/ml-pipeline-cpu` (2026-05-08).
